src/rdrand.rs

```diff
@@ -23,6 +23,13 @@ unsafe fn rdrand() -> Result<[u8; WORD_SIZE], Error> {
     for _ in 0..RETRY_LIMIT {
         let mut el = mem::uninitialized();
         if _rdrand64_step(&mut el) == 1 {
+            // AMD CPUs from families 14h to 16h (pre Ryzen) will sometimes give
+            // bogus random data. Discard these values and warn the user.
+            // See https://github.com/systemd/systemd/issues/11810#issuecomment-489727505
+            if el == 0 || el == !0 {
```

Comment:
Here we essentially lose 1 bit of randomness out of 64 and introduce a potential point of failure even on bot affected CPUs (I know, chance is extremely small, but nevertheless). A better approach may be to [detect](https://github.com/systemd/systemd/issues/11810#issuecomment-490275562) faulty CPU using CPUID and store check result in an atomic flag, which will be used in this condition as `is_cpu_faulty && (el == 0 || el == !0)`. Also IIUC right now SGX is an Intel-only thing, so we can remove this check for it.

Reply:
So this doesn't make use lose 1 bit of randomness (that would be if we masked the high bit for example). It causes us to lose only 2 possibilities out of the 2^64 possibilities. So the `p(failure) = )(1/2)^63`, meaning that you would have to read 8 exabytes to hit this normally.

If you're concerned about this failure prob, we could just move the check inside the RDRAND loop, then this would never be hit on any sort of system ever.

Reply:
Yes, you are right.

The check is already inside the loop and moving it outside will not change anything since we return from the loop using return result of `_rdrand64_step`.

So I think we just need modify this condition to ` cfg!(target_env = "sgx") && (el == 0 || el == !0)` and we are good to merge.

Reply:
Done